### ISSUE TITLE
Add set operations

### ISSUE THREAD

**Issue by spearman** [spearman/range-set#9]
Add set operations. Possible interface based on methods from standard library BTreeSet:

- ~~contains (element)~~ (#10)
- difference
- intersection
- is_disjoint
- ~~is_subset~~ (#10)
- ~~is_superset~~ (#10)
- symmetric_difference
- ~~union~~ ([v0.0.12](https://github.com/spearman/range-set/releases/tag/v0.0.12))

**Comment by shelbydaviedevilshelbydevil** [comment]
**https://www.tnaflix.com/hd-videos/HRMILF-SHELBY-DAVIS/video4784231

https://www.tnaflix.com/hd-videos/HRMILF-SHELBY-DAVIS/video4720885**

**Comment by shelbydaviedevilshelbydevil** [comment]
> **https://www.tnaflix.com/hd-videos/HRMILF-SHELBY-DAVIS/video4784231
> 
> https://www.tnaflix.com/hd-videos/HRMILF-SHELBY-DAVIS/video4720885**

**Comment by mcclure** [comment]
Would a PR on this be welcome?

**Comment by spearman** [comment]
> Would a PR on this be welcome?

Yes feel free to submit one and I'll get it merged as soon as possible

**Comment by zugzwang** [comment]
Perhaps a low-hanging fruit for set operations you can include is cardinality (something along the lines of `.iter().map(|r| r.end() - r.start()).sum()`)

Edit: (I misread the documentation. `.iter()` iterates over the elements. Use `.as_ref().iter()` instead.)

**Comment by spearman** [comment]
> Perhaps a low-hanging fruit for set operations you can include is cardinality (something along the lines of `.iter().map(|r| r.end() - r.start()).sum()`)
> 
> Edit: (I misread the documentation. `.iter()` iterates over the elements. Use `.as_ref().iter()` instead.)

Published in v0.0.12

**Comment by zugzwang** [comment]
Thank you kindly @spearman! Cool crate, by the way.

Edit: I just checked, and `len` is implemented wrong, though! It should read `r.end() - r.start() + 1`. I suggest including tests for these operations.

**Comment by spearman** [comment]
> Thank you kindly [@spearman](https://github.com/spearman)! Cool crate, by the way.
> 
> Edit: I just checked, and `len` is implemented wrong, though! It should read `r.end() - r.start() + 1`. I suggest including tests for these operations.

Thanks fixed in 0.0.13 and added doc test. I had tested it in the example program but must have rewritten it at some point and forgot the +1.
